File: leetcode_75/sliding_window/maximum_number_of_vowels_in_a_substring_of_given_length_1456.java
package leetcode_75.sliding_window;

import java.util.Set;

class Solution {
    public int maxVowels(String s, int k) {
        Set<Character> vowel = Set.of('a', 'e', 'i', 'o', 'u');

        int left = 0, right = k - 1, temp = 0;

        for (int i = left; i <= right; i++) {
            if (vowel.contains(s.charAt(i))) {
                temp++;
            }
        }

        int m_vowel = temp;

        while (right < s.length() - 1) {
            if (vowel.contains(s.charAt(left))) {
                temp--;
            }
            left++;
            right++;
            if (vowel.contains(s.charAt(right))) {
                temp++;
            }
            m_vowel = Math.max(m_vowel, temp);
        }
        return m_vowel;
    }
}

File: leetcode_75/sliding_window/maximum_number_of_vowels_in_a_substring_of_given_length_1456.py
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowel = set(('a', 'e', 'i', 'o', 'u'))

        left, right = 0, k-1
        temp = sum(1 for i in range(left, right+1) if s[i] in vowel)
        m_vowel = temp

        while right < len(s) -1:
            if s[left] in vowel:
                temp -= 1
            left += 1

            right +=1
            if s[right] in vowel:
                temp += 1

            m_vowel = max(m_vowel, temp)
        return m_vowel
# ------------------------------------------------------------
# 這超時了
    def maxVowels(self, s: str, k: int) -> int:
        vowel = set(('a', 'e', 'i', 'o', 'u'))
        ll = list(s)

        left, right, m_vowel = 0, k, 0
        # temp = []

        while right < len(ll)+1:
            temp = 0
            for i in range(left, right):
                if ll[i] in vowel:
                    temp +=1
            left += 1
            right +=1
            m_vowel = max(m_vowel, temp)
        return m_vowel